### PR TITLE
RavenDB-5435 added the ability to compact an index

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 1]
```diff
@@ -1695,6 +1743,66 @@ public bool CanContinueBatch(IndexingStatsScope stats)
...
+
+            if (_environment.Options.IncrementalBackupEnabled)
+                throw new InvalidOperationException($"Index '{Name} ({IndexId})' cannot be compacted because incremental backup is enabled.");
+
+            if (Configuration.RunInMemory)
+                throw new InvalidOperationException($"Index '{Name} ({IndexId})' cannot be compacted because it runs in memory.");
+
+            lock (_locker)
+            {
+                _isCompactionInProgress = true;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs]: This isn't thread safe.
what is going to happen to queries that are currently running? Or indexing that is currently using?
You are pulling the carpet on them by disposing the environment, and then making the memory they use invalid.

You need to add a reader / writer lock here to ensure that we are the only one accessing it.

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ppekrol** [comment]
I will change approach.

**Comment by ppekrol** [comment]
Draining queries before compaction now.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -1756,11 +1758,22 @@ public IOperationResult Compact(Action<IOperationProgress> onProgress)
...
                 _isCompactionInProgress = true;
+
+                var progress = new IndexCompactionProgress();
+                var currentlyRunningQueriesCount = CurrentlyRunningQueries.Count;
+                while (currentlyRunningQueriesCount > 0)
+                {
+                    progress.Message = $"Waiting for queries to end. {currentlyRunningQueriesCount} queries are still processing.";
+                    onProgress.Invoke(progress);
+
+                    Thread.Sleep(TimeSpan.FromSeconds(1));
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:1784]: You are sleeping while holding the lock, that isn't a very good idea.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 3]
```diff
@@ -1756,11 +1758,22 @@ public IOperationResult Compact(Action<IOperationProgress> onProgress)
             lock (_locker)
             {
                 _isCompactionInProgress = true;
+
+                var progress = new IndexCompactionProgress();
+                var currentlyRunningQueriesCount = CurrentlyRunningQueries.Count;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:1778]: There is a race here, a user can check the `_isCompactionInProgress` in `AssertIndexState`, see it is valid, and then start querying, and we check if there are running queries before that. 
You can't rely on this

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 4]
```diff
@@ -1756,11 +1758,22 @@ public IOperationResult Compact(Action<IOperationProgress> onProgress)
             lock (_locker)
             {
                 _isCompactionInProgress = true;
+
+                var progress = new IndexCompactionProgress();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs:1777]: Who is this reporting to? It looks like this goes into the void.

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs:1777]: I think you meant to add this to the operations?

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Index.cs:1777]: It is added to the operations, check https://github.com/ppekrol/ravendb/blob/v4.0/src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs#L28

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Index.cs:1784]: Fixed

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 3 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Index.cs:1778]: Fixed

**Comment by ayende** [comment]
User RW implemetnation for this.

**Comment by ppekrol** [comment]
Ok, note that RW is not async/await compatible IIRC

**Comment by ayende** [comment]
Yes, I handled that by exiting the lock before the await

**Pull request closed without merging**